### PR TITLE
Inputty depends_on support for Python

### LINKED ISSUE #5642: depends_on only works with a Resource/CustomResource and not Output[CustomResource]
## Expected Behavior
Trying to use the get_resource method of a Helm3 Chart to extract a created resource for a depends_on. 

## Current Behavior
When I use that extracted Output[CustomResource] it receive the error `Exception: 'depends_on' was passed a value that was not a Resource.` Most likely the check on depends_on needs to be fixed/extended to work with Output[Resource/CustoResource] types. 

## Steps to Reproduce
```
cert_manager_deployment = cert_manager_chart.get_resource(group_version_kind="apps/v1/Deployment",
                                                          name="cert-manager",
                                                          namespace="cert-manager")
```
Use that as a depends_on resource.

https://github.com/pulumi/pulumi/blame/5cef84f0366f264baf51c322d1d22b6124fdf236/sdk/python/lib/pulumi/resource.py#L450-L454

### PR DESCRIPTION
**Pull request by t0yv0**
# Description

Allows passing `Input[Resource]` into `depends_on` option of a `Resource`. The implied promises will be awaited during RPC.

- [x] tests
- [x] check how Node handles unknown=True outputs passed into `depends_on`
- [x] decide if we want to try to recover known values passed into `depends_on` if some are unknown 

Fixes #5642  (for Python, Go and C# remaining)

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: sdk/python/lib/pulumi/output.py [hunk 1]
```diff
@@ -440,6 +440,23 @@ def contains_unknowns(val: Any) -> bool:
     return rpc.contains_unknowns(val)
 
 
+def _is_prompt(value: Input[T]) -> bool:
```
Comment (t0yv0) [inline sdk/python/lib/pulumi/output.py:443]: Am I missing something similar here in the SDK? This was a bit non-trivial to write as I'm wondering if I understand the `is_input_type` clause and should I be checking anything else.

File: sdk/python/lib/pulumi/resource.py [hunk 2]
```diff
@@ -586,6 +589,26 @@ def _merge_lists(dest, source):
     return dest + source
 
 
+def _append_input_lists(a: Optional['Input[List[T]]'],
```
Comment (t0yv0) [inline sdk/python/lib/pulumi/resource.py:592]: What I find the SDK lacking is something along these lines:

```
map_input : Callable[[A],B] -> Input[A] -> Input[B]
map2_inputs : Callable[[A,B],C] -> Input[A] -> Input[B] -> Input[C]
```

This functions would try to preserve promptness, dropping to Output only if really necessary (which is what we are trying to do here). Could use it to zip the inputty lists then:

```
map2_inputs((lambda xs, ys: xs + ys), a , b)
```

File: sdk/python/lib/pulumi/runtime/resource.py [hunk 3]
```diff
@@ -140,7 +139,8 @@ async def prepare_resource(res: 'Resource',
         for dep in deps:
             urn = await dep.urn.future()
             urns.add(urn)
-            dependencies.add(urn)
+            if urn:
```
Comment (t0yv0) [inline sdk/python/lib/pulumi/runtime/resource.py:142]: Something weird was going on here in the SDK before the change, why `Set[Optional[str]]` is being returned for URN-sets? DO we care about a `None` tracked in the URN sets? Probably not, need to see if can refactor.

File: sdk/python/lib/pulumi/runtime/resource.py [hunk 4]
```diff
@@ -657,3 +659,30 @@ def convert_providers(
         result[p.package] = p
 
     return result
+
+
+async def _resolve_depends_on_urns(options: 'ResourceOptions') -> List[str]:
+    """
+    Resolves the set of all dependnet resources implied by
```
Comment (t0yv0) [inline sdk/python/lib/pulumi/runtime/resource.py:666]: Typo

File: sdk/python/lib/pulumi/runtime/mocks.py [hunk 5]
```diff
@@ -31,9 +32,12 @@
 
 
 def test(fn):
+
+    @functools.wraps(fn)
```
Comment (t0yv0) [inline sdk/python/lib/pulumi/runtime/mocks.py:36]: The effect of this decorator is such that the wrapped function preserves the same signature (such as argument names) as the original. This ends up being important for name-based fixture matching functionality of pytest. So now when I write

```
@pulumi.runtime.test
def test_foo(my_fixture):
   ...
```

It works, which it didn't previously.

**Comment by t0yv0** [comment]
Node avoids eliding deps if some deps in the list are unknown. Added a test and matched the behavior here.

File: sdk/python/lib/pulumi/output.py [hunk 6]
```diff
@@ -297,6 +297,35 @@ def from_input(val: Input[T_co]) -> 'Output[T_co]':
         value_fut.set_result(val)
         return Output(set(), value_fut, is_known_fut, is_secret_fut)
 
+    @staticmethod
+    def _from_input_shallow(val: Input[T]) -> 'Output[T]':
+        """
+        Like `from_output`, but does not recur deeply. Instead, checks if `val` is an `Output` value
```
Comment (justinvp) [inline sdk/python/lib/pulumi/output.py:303]: Did you mean:

```suggestion
        Like `from_input`, but does not recur deeply. Instead, checks if `val` is an `Output` value
```

?

File: sdk/python/lib/pulumi/output.py (hunk 1 above)
Comment (justinvp) [reply sdk/python/lib/pulumi/output.py:477]: I don't think we should be checking `is_input_type` here -- all that does is check to see if it's a class that has the `@input_type` decorator applied. It's unrelated to `Input[T]`.

File: sdk/python/lib/pulumi/runtime/resource.py (hunk 3 above)
Comment (justinvp) [reply sdk/python/lib/pulumi/runtime/resource.py:145]: > DO we care about a `None` tracked in the URN sets?

I don't think so. Do you know why `None` was being added to the set?

File: sdk/python/lib/test/test_resource.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+class MinimalMocks(pulumi.runtime.Mocks):
+
+    def new_resource(self, args: pulumi.runtime.MockResourceArgs):
+        return [args.name + '_id', args.inputs]
+
+    def call(self, args: pulumi.runtime.MockCallArgs):
+        return {}
+
+
+class DependencyTrackingMonitorWrapper(object):
```
Comment (justinvp) [inline sdk/python/lib/test/test_resource.py:98]: Nit: This shouldn't be necessary on the versions of Python we support (3.6+):

```suggestion
class DependencyTrackingMonitorWrapper:
```

File: sdk/python/lib/test/test_resource.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+import asyncio
+import pytest
+
+from pulumi.runtime import settings, mocks
+from pulumi.runtime.proto import resource_pb2
+import pulumi
+
+
+@pulumi.runtime.test
+def test_depends_on_accepts_outputs(dep_tracker):
```
Comment (justinvp) [inline sdk/python/lib/test/test_resource.py:25]: Should we test more permutations of acceptable values for `depends_on` similar to the test for [Node.js](https://github.com/pulumi/pulumi/pull/2198/files#diff-2af4e445f769dffd2cac643d6ec6b50d91fe04731c19eae1267a7fbfefeb4b58)?

https://github.com/pulumi/pulumi/blob/30311405ca1a48f3b3d704981eb5c87e422c2ca9/sdk/nodejs/tests/runtime/langhost/cases/008.ten_depends_on_resources/index.js#L13-L38

**Review by justinvp — APPROVED** [review]
LGTM otherwise

File: sdk/python/lib/test/test_resource.py (hunk 8 above)
Comment (t0yv0) [reply sdk/python/lib/test/test_resource.py:25]: Didn't see that test, that's pretty helpful, looking.

File: sdk/python/lib/pulumi/output.py (hunk 1 above)
Comment (t0yv0) [reply sdk/python/lib/pulumi/output.py:477]: Agreed, after reflecting about it and writing my "haskeller's impression" doc, this check is not needed here. The entire thing makes a bit more sense now.

File: sdk/python/lib/test/test_resource.py (hunk 8 above)
Comment (t0yv0) [reply sdk/python/lib/test/test_resource.py:28]: This is great, this discovered a gap wrt supporting "non-listy" versions of the API, they are certainly convenient, will implement those.

File: sdk/python/lib/pulumi/runtime/resource.py (hunk 3 above)
Comment (t0yv0) [reply sdk/python/lib/pulumi/runtime/resource.py:145]: I've poked around and it sounds like Optional[str] sequence is passed to proto layer; it looks harmless I don't think there's any semantics attached to this, we can probably tighten to str.

**Comment by t0yv0** [comment]
@justinvp mind having another quick look that the newly introduced map, map2 combinators look okay? I'm not using them in positions requiring reliability yet but might do so in the future. Tyvm.

**Review by justinvp — APPROVED** [review]

**Comment by t0yv0** [comment]
I was really unlucky with flaky tests here, had to rerun the CI what feels like lots of times.. Very unfortunate.

**Pull request merged**